### DISCUSSION TITLE
USB reemulation

### DISCUSSION THREAD

**Discussion by 203Null** [hathach/tinyusb#1904] [Ideas]
# Below are the original github issue. Please ignore.

### Related area

Class Driver

### Hardware specification

Any

### Is your feature request related to a problem?

No problem, just ideas.

Since all interface is determined upon connection, dynamic configuration is not possible (I think?). However, if the device can emulate a hub and can then maybe we can connect more devices to this hub to do dynamic configuration on the fly.

Not entirely sure how practical this but just throwing an idea for a potential feature,

### Describe the solution you'd like

hub class for device class

### I have checked existing issues, dicussion and documentation

- [x] I confirm I have checked existing issues, dicussion and documentation.

**Comment by hathach** [comment]
hub device is too difficult and will be impossible for most MCUs since hub mostly work in physical layer. I don't see why we should spend that much of effort to implement this. Most (95%) of people won't use it, forcing an re-enumerate is much better/faster.

**Reply by 203Null** [reply]
I see, Thank you.

If I understand the [dynamic_configuration](https://github.com/hathach/tinyusb/tree/master/examples/device/dynamic_configuration) example correctly. It is switching descriptors based on if the button is pressed during emulation. 

How can I trigger re-emulation without physically reconnecting the USB? Via hardware disconnect the USB (like a pull-up on the USB D+)? Or it can be done via TinyUSB? If I remember correctly deinit wasn't implemented yet.

**Reply by hathach** [reply]
yeah, disable/reenable D+ pull up, on some port it is implemented with tud_disconnect()/connect() API.

**Reply by 203Null** [reply]
Thank you! I will give it a try.
